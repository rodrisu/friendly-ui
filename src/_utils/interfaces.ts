export type CommonFieldsProps = {
  id?: string
  name: string
  className?: string
  value: string
  disabled?: boolean
  autoFocus?: boolean
  required?: boolean
}

export type A11yProps = {
  id?: string
  role?: string
  title?: string
  tabIndex?: number
  'aria-label'?: string
  'aria-labelledby'?: string
  'aria-describedby'?: string
  'aria-controls'?: string
}

type A11yKeys = keyof A11yProps

export function pickA11yProps<T extends A11yProps>(source: T): Pick<T, A11yKeys> {
  const initialValue = {} as Pick<T, A11yKeys>
  const keys: Array<A11yKeys> = [
    'id',
    'role',
    'title',
    'tabIndex',
    'aria-label',
    'aria-labelledby',
    'aria-describedby',
    'aria-controls',
  ]

  return keys.reduce(
    (acc, key) => ({
      ...acc,
      [key]: source[key],
    }),
    initialValue,
  )
}
