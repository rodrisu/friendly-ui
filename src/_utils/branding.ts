type Color = {
  [propName: string]: string
}

export const color: Color = {
  white: '#FFF',
  gray: '#DDD',
  lightGray: '#EDEDED',
  midnightGreen: '#054752',
  midnightGreenDetail: '#3D5C62',
  lightMidnightGreen: '#708C91',
  blue: '#00AFF5',
  darkBlue: '#008FC1',
  boostBlue: '#2ED1FF',
  green: '#5DD167',
  orange: '#F78B00',
  yellow: '#FFCA0C',
  red: '#F53F5B',
  lightRed: '#FDD8DE',
  facebookBrand: '#1877F2',
  vkBrand: '#2787F5',
  blaBlaCarDailyBrand: '#00D8A7',
  blaBlaCarDailyBrandText: '#30004B',
  tapHighlight: 'rgba(221, 221, 221, .4)', // gray, 40%
  lightBlue: '#ADECFE',
}

export const gradientColors = {
  blueGradient: `linear-gradient(80.19deg, ${color.boostBlue} 4.92%, ${color.blue} 95.98%)`,
}

type Font = {
  [key: string]: {
    size: string
    lineHeight: string
  }
}

type defaultBranding = {
  [key: string]: string
}

export const CDN_FONTS_BASE_URL = 'https://cdn.blablacar.com/kairos/assets/build/fonts'

export const font: Font = {
  s: {
    size: '13px',
    lineHeight: '16px',
  },
  base: {
    size: '16px',
    lineHeight: '20px',
  },
  m: {
    size: '18px',
    lineHeight: '20px',
  },
  l: {
    size: '22px',
    lineHeight: '24px',
  },
  xl: {
    size: '30px',
    lineHeight: String(Math.floor((32 / 30) * 100) / 100),
  },
  brand: {
    size: '56px',
    lineHeight: '56px',
  },
  xxl: {
    size: '82px',
    lineHeight: '82px',
  },
}

export const fontWeight: defaultBranding = {
  regular: '400',
  medium: '500',
}

export const space: defaultBranding = {
  none: '0',
  xs: '2px',
  s: '4px',
  m: '8px',
  l: '16px',
  xl: '24px',
  xxl: '48px',
}

export const horizontalSpace: defaultBranding = {
  inner: '8px',
  outer: '16px',
  global: '24px',
}

export const radius: defaultBranding = {
  none: '0',
  xs: '2px',
  s: '4px',
  m: '8px',
  l: '16px',
  xl: '24px',
}

// In milliseconds.
export enum TransitionDuration {
  BASE = 200,
  FAST = 500,
}

export const transition = {
  duration: {
    base: `${TransitionDuration.BASE}ms`,
    fast: `${TransitionDuration.FAST}ms`,
  },
  easing: {
    default: 'ease-in',
  },
  callbackDelay: 1000,
}

export const delay = {
  timeout: {
    base: '400ms',
  },
  interval: {
    base: '100ms',
  },
}

export const componentSizes = {
  timeWidth: '48px',
  buttonIconSize: '48px',
  bulletSize: '10px',
  bulletSizeSmall: '8px',
  bulletSizeMap: '18px',
  bulletSizeSearch: '18px',
  roadWidth: '4px',
  smallSectionWidth: '662px',
  largeSectionWidth: '1016px',
  modalTopPadding: '70px',
  searchOverlayWidth: '375px',
  headerHeight: {
    small: '56px',
    large: '72px',
  },
  headerWidth: {
    small: '100%',
    large: '1280px',
  },
  drawerWidth: {
    small: '100%',
    large: '350px',
  },
}

export const modalSize = {
  s: '400px',
  m: '662px',
  l: '928px',
}

export const shadow: defaultBranding = {
  default: '0 2px 4px rgba(0, 0, 0, .3)',
  icon: '0 0 2px rgba(0, 0, 0, .3)',
  card: '0 1pt 4pt rgba(0, 0, 0, 0.16), 0 2pt 8pt rgba(0, 0, 0, 0.08)',
  cardHover: '0 2pt 8pt rgba(0, 0, 0, 0.08), 0 2pt 16pt rgba(0, 0, 0, 0.08)',
  searchForm: '0 1pt 4pt rgba(0, 0, 0, 0.16), 0 2pt 8pt rgba(0, 0, 0, 0.08)',
  slideSection: '0 4px 4px rgba(0, 0, 0, .25)',
}

export const inputBorderSize = {
  default: '1px',
  focus: '3px',
}

const largeBreakPointThreshold = '800px'
const smallBreakPointThreshold = '799px'
export const responsiveBreakpoints = {
  small: smallBreakPointThreshold,
  isMediaLarge: `min-width: ${largeBreakPointThreshold}`,
  isMediaSmall: `max-width: ${smallBreakPointThreshold}`,
}

export const selectHeight = '52px'

export const pxToInteger = (size: string): number => parseInt(size.replace('px', ''), 10)

export const branding = {
  color,
  font,
  fontWeight,
  space,
  radius,
  shadow,
}
