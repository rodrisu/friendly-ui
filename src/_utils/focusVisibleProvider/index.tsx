import React, { createContext, ReactNode, useEffect, useState } from 'react'

import { KEYS, KEYS_TRIGGERING_KEYBOARD_NAVIGATION } from '../keycodes'

// A React hook based on: https://github.com/WICG/focus-visible

const pointerMoveEventList = [
  'mousedown',
  'mouseup',
  'pointermove',
  'pointerdown',
  'pointerup',
  'touchmove',
  'touchstart',
  'touchend',
]

const pointerDownEventList = ['mousedown', 'pointerdown', 'touchstart']

export const FOCUS_VISIBLE_CSS_CLASS = 'focus-visible'

export type FocusVisibleProviderProps = {
  children: ReactNode
  // Provide a global css selector for legacy components.
  setGlobalClassName?: boolean
}

export const FocusVisibleContext = createContext(false)

export const FocusVisibleProvider = ({
  children,
  setGlobalClassName,
}: FocusVisibleProviderProps) => {
  // When the provider first loads, assume the user is in pointer modality.
  const [hadKeyboardEvent, setHadKeyboardEvent] = useState(false)

  if (setGlobalClassName) {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    useEffect(() => {
      // Add a 'focus-visible' CSS class on the <body> element.
      document.body.classList.toggle(FOCUS_VISIBLE_CSS_CLASS, hadKeyboardEvent)
    }, [hadKeyboardEvent])
  }

  useEffect(() => {
    let lastClientX: Number
    let lastClientY: Number

    const onPointerDown = () => {
      setHadKeyboardEvent(false)
    }

    const onInitialPointerMove = () => {
      setHadKeyboardEvent(false)
    }

    const getLastMouseMove = (e: MouseEvent) => {
      // Ensure the mouse has actually moved (Safari)
      // https://transitory.technology/mouse-move-in-safari/
      if (lastClientX === e.clientX && lastClientY === e.clientY) {
        return
      }

      lastClientX = e.clientX
      lastClientY = e.clientY
    }

    const addInitialPointerMoveListeners = () => {
      document.addEventListener('mousemove', (event: MouseEvent) => {
        getLastMouseMove(event)
        // eslint-disable-next-line no-unused-expressions
        onInitialPointerMove
      })

      pointerMoveEventList.forEach(e => document.addEventListener(e, onInitialPointerMove))
    }

    const removeInitialPointerMoveListeners = () => {
      document.removeEventListener('mousemove', (event: MouseEvent) => {
        getLastMouseMove(event)
        // eslint-disable-next-line no-unused-expressions
        onInitialPointerMove
      })

      pointerMoveEventList.forEach(e => document.removeEventListener(e, onInitialPointerMove))
    }

    const onKeyDown = (e: KeyboardEvent) => {
      const element = e.target as HTMLInputElement
      const isTypingArea =
        element.tagName === 'TEXTAREA' || (element.tagName === 'INPUT' && element.type === 'text')
      // Remove Spacebar and Enter keys in case of text editing
      const keysList = isTypingArea
        ? KEYS_TRIGGERING_KEYBOARD_NAVIGATION.filter(
            key => key !== KEYS.SPACEBAR && key !== KEYS.ENTER,
          )
        : KEYS_TRIGGERING_KEYBOARD_NAVIGATION

      if (keysList.includes(e.key)) {
        setHadKeyboardEvent(true)
      }
    }

    // For some kinds of state, we are interested in changes at the global
    // scope only. Global pointer input and global key presses change
    // should affect the state at every scope.
    document.addEventListener('keydown', onKeyDown, true)
    pointerDownEventList.forEach(e => document.addEventListener(e, onPointerDown, true))
    addInitialPointerMoveListeners()

    return () => {
      document.removeEventListener('keydown', onKeyDown, true)
      pointerDownEventList.forEach(e => document.removeEventListener(e, onPointerDown, true))
      removeInitialPointerMoveListeners()
    }
  }, [])

  return (
    <FocusVisibleContext.Provider value={hadKeyboardEvent}>{children}</FocusVisibleContext.Provider>
  )
}
