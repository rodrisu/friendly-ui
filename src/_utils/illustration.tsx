import React from 'react'
import isEmpty from 'lodash.isempty'
import styled from 'styled-components'

export type IllustrationProps = Readonly<{
  title?: string
}>

type BaseIllustrationProps = IllustrationProps &
  Readonly<{
    children: React.ReactNode
    viewBox?: string
  }>

const StyledIllustration = styled.svg`
  max-height: 33vh;
  width: 100%;
  object-fit: contain;
`

export const Illustration = ({
  children,
  title = '',
  viewBox = '0 0 375 204',
}: BaseIllustrationProps): JSX.Element => (
  <StyledIllustration
    viewBox={viewBox}
    fill="none"
    xmlns="http://www.w3.org/2000/svg"
    aria-hidden={isEmpty(title)}
  >
    {title && <title>{title}</title>}
    {children}
  </StyledIllustration>
)
