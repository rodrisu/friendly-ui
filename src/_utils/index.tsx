import React, { Fragment } from 'react'
import { canUseDOM } from 'exenv'

const DEFAULT_CLASS_PREFIX = 'kirk'

export const prefix = (modifiers: Classcat.ClassObject, baseClass?: string): string => {
  const mods = Object.keys(modifiers).filter(elem => modifiers[elem])

  let classPrefix = DEFAULT_CLASS_PREFIX

  if (baseClass != null) {
    classPrefix = `${classPrefix}-${baseClass}`
  }

  return mods.map(modifier => `${classPrefix}-${modifier}`).join(' ')
}

export const isTouchEventsAvailable = () =>
  canUseDOM && 'ontouchstart' in window && 'ontouchend' in window

export const replaceNewLineWithBR = (str: string): React.ReactNode =>
  str
    .split('\n')
    .map(line => <Fragment>{line}</Fragment>)
    .reduce((acc, curr) => (
      <Fragment>
        {acc}
        <br />
        {curr}
      </Fragment>
    ))
