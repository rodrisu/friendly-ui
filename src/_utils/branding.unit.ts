import { pxToInteger } from './branding'

it('should convert a string px value to integer', () => {
  expect(pxToInteger('10px')).toEqual(10)
})
