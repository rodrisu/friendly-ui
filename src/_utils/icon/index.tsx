import styled from 'styled-components'

import { color, shadow } from '../branding'
import { BaseIcon } from './BaseIcon'

const StyledBaseIcon = styled(BaseIcon)`
  fill: ${props => props.iconColor ?? color.lightMidnightGreen};

  &,
  & rect,
  & circle,
  & line,
  & polyline,
  & path,
  & g {
    stroke: ${props => (props.isDisabled ? color.gray : '')};
  }

  &.kirk-icon-wrapper {
    display: inline-block;
    position: relative;
  }

  &.kirk-icon-shadowed {
    border-radius: 50%;
    box-shadow: ${shadow.icon};
  }

  & .kirk-icon-badge {
    position: absolute;
    bottom: 12px;
    left: 12px;
  }
`

export * from './BaseIcon'

export { StyledBaseIcon as BaseIcon }
