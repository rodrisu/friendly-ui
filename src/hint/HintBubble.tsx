import React from 'react'
import cc from 'classcat'
import styled from 'styled-components'

import { color, fontWeight, horizontalSpace, radius, shadow, space } from '../_utils/branding'
import { A11yProps, pickA11yProps } from '../_utils/interfaces'
import { Button, ButtonStatus } from '../button'
import { CrossIcon } from '../icon/crossIcon'

export enum HintBubblePosition {
  ABOVE = 'above',
  BELOW = 'below',
}

export type HintBubbleProps = A11yProps &
  Readonly<{
    mainTitle: string
    className?: string
    onClose: () => void
    closeButtonTitle?: string
    description?: string
    position?: HintBubblePosition
  }>

const HintBubble = (props: HintBubbleProps): JSX.Element => {
  const a11yProps = pickA11yProps<HintBubbleProps>(props)
  const { mainTitle, description, closeButtonTitle, position, className, onClose } = props
  return (
    <aside className={cc([className, `bubble-arrow--${position}`])} {...a11yProps}>
      <p>
        <strong>{mainTitle}</strong>
        {description}
      </p>
      <Button
        onClick={onClose}
        title={closeButtonTitle}
        status={ButtonStatus.UNSTYLED}
        aria-controls={a11yProps.id}
      >
        <CrossIcon size={18} iconColor={color.white} />
      </Button>
    </aside>
  )
}

const StyledHintBubble = styled(HintBubble)`
  & {
    position: relative;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: flex-start;
    padding: ${space.l};
    background-color: ${color.midnightGreen};
    color: ${color.white};
    border-radius: ${radius.l};
    box-shadow: ${shadow.icon};
    /* Normalization */
    margin-left: ${horizontalSpace.global};
    margin-right: ${horizontalSpace.global};
  }

  & strong {
    display: block;
    font-weight: ${fontWeight.medium};
    padding-bottom: ${space.s};
  }

  & p {
    margin: 0;
    padding: 0;
    flex: 1;
  }

  & .kirk-button {
    padding: 0 0 0 ${space.m};
    min-height: 0;
    height: auto;
  }

  & .kirk-button svg {
    margin: 0;
  }

  &.bubble-arrow--above::after {
    content: ' ';
    position: absolute;
    top: 100%;
    left: 50%;
    margin-left: -4px;
    width: 0;
    height: 0;
    border-left: 8px solid transparent;
    border-right: 8px solid transparent;
    border-top: 8px solid ${color.midnightGreen};
  }

  &.bubble-arrow--below::after {
    content: ' ';
    position: absolute;
    bottom: 100%;
    left: 50%;
    margin-left: -4px;
    width: 0;
    height: 0;
    border-left: 8px solid transparent;
    border-right: 8px solid transparent;
    border-bottom: 8px solid ${color.midnightGreen};
  }
`
export { StyledHintBubble as HintBubble }
