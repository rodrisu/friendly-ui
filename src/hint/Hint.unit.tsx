import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

import { A11yProps } from '../_utils/interfaces'
import { HintBubble } from './HintBubble'
import { Hint, HintBubblePosition } from './index'

const testChild = (a11yAttrs: A11yProps): React.ReactNode => <p {...a11yAttrs}>I have an hint.</p>

const defaultProps = {
  mainTitle: 'Hint Title',
  closeButtonTitle: 'Close',
  description: 'Hint Description',
}

describe('Hint', () => {
  it('Default rendering (above)', () => {
    const hint = renderer.create(<Hint {...defaultProps}>{testChild}</Hint>).toJSON()
    expect(hint).toMatchSnapshot()
  })
  it('Default rendering (below)', () => {
    const hint = renderer
      .create(
        <Hint {...defaultProps} position={HintBubblePosition.BELOW}>
          {testChild}
        </Hint>,
      )
      .toJSON()
    expect(hint).toMatchSnapshot()
  })
  it('Should render a HintBubble', () => {
    const hint = shallow(<Hint {...defaultProps}>{testChild}</Hint>)
    expect(hint.find(HintBubble)).toHaveLength(1)
  })
  it('Should not render a HintBubble if prop hidden is set to true', () => {
    const hint = shallow(
      <Hint {...defaultProps} hidden>
        {testChild}
      </Hint>,
    )
    expect(hint.find(HintBubble)).toHaveLength(0)
  })
})
