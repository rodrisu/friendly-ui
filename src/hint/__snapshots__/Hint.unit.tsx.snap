// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hint Default rendering (above) 1`] = `
.c4 {
  fill: #FFF;
}

.c4.kirk-icon-wrapper {
  display: inline-block;
  position: relative;
}

.c4.kirk-icon-shadowed {
  border-radius: 50%;
  box-shadow: 0 0 2px rgba(0,0,0,.3);
}

.c4 .kirk-icon-badge {
  position: absolute;
  bottom: 12px;
  left: 12px;
}

.c3 {
  position: relative;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding: 0 24px;
  border: 1px solid transparent;
  border-radius: 48px;
  font-size: 16px;
  line-height: 1;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  font-family: inherit;
  height: 48px;
  min-width: 48px;
  max-width: 100%;
  overflow: hidden;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: max-width 500ms ease-in, background-color 200ms ease-in;
  transition: max-width 500ms ease-in, background-color 200ms ease-in;
  white-space: nowrap;
}

.c3 svg,
.c3 img {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c3 svg {
  margin-right: 8px;
}

.c3.kirk-button-bubble svg {
  margin: 0;
}

.c3.kirk-itemChoice {
  border-radius: 0;
  overflow: visible;
  width: 100%;
}

.c3:hover,
.c3:active {
  outline: 0;
}

.c3.kirk-button-unstyled {
  background-color: transparent;
  border-radius: inherit;
  color: #00AFF5;
  padding: inherit;
  display: inherit;
  line-height: inherit;
  font-size: inherit;
  text-align: left;
  min-width: inherit;
  min-height: inherit;
  height: auto;
  white-space: normal;
}

.c3[disabled] {
  opacity: 0.5;
  cursor: default;
}

.c3.kirk-button-unstyled[disabled] {
  opacity: 1;
}

.c3.kirk-button-primary {
  background-color: #00AFF5;
  color: #FFF;
}

.c3.kirk-button-shadowed {
  border: 2px solid #FFF;
  box-shadow: 0 2px 4px rgba(0,0,0,.3);
  box-sizing: border-box;
}

.c3.kirk-button-primary:hover,
.c3.kirk-button-primary:focus,
.c3.kirk-button-primary:active {
  background-color: #008FC1;
}

.c3.kirk-button-secondary {
  background-color: #FFF;
  border-color: #DDD;
  color: #00AFF5;
}

.c3.kirk-button-secondary.kirk-button-shadowed {
  border: none;
}

.c3.kirk-button-secondary:hover,
.c3.kirk-button-secondary:focus,
.c3.kirk-button-secondary:active {
  background-color: #DDD;
}

.c3.kirk-button-tertiary {
  background-color: #FFF;
  border-color: transparent;
  color: #00AFF5;
}

.c3.kirk-button-tertiary:hover,
.c3.kirk-button-tertiary:focus,
.c3.kirk-button-tertiary:active {
  background-color: #DDD;
}

.c3.kirk-button-loading {
  background-color: transparent;
  max-width: initial;
  border: 0;
}

.c3.kirk-button-loading:hover,
.c3.kirk-button-loading:focus,
.c3.kirk-button-loading:active {
  background-color: transparent;
}

.c3.kirk-button-loading[disabled],
.c3.kirk-button-checked[disabled] {
  opacity: 1;
}

.c3 svg.kirk-button-loader {
  position: absolute;
  top: 0;
  left: 50%;
  margin-left: -24px;
}

.c3.kirk-button-checked {
  border: 0;
}

.c3.kirk-button-warning {
  background-color: #F53F5B;
  color: #FFF;
}

.c3.kirk-button-bubble {
  display: inline-block;
  text-align: center;
  padding: 0;
  line-height: 0;
  width: 48px;
  height: 48px;
}

.c3.kirk-button-bubble.kirk-button-shadowed {
  width: calc(48px - 2px * 2);
  height: calc(48px - 2px * 2);
}

.c3.kirk-button-facebook {
  background-color: #1877F2;
  color: #FFF;
}

.c3.kirk-button-vkontakte {
  background-color: #2787F5;
  color: #FFF;
}

.c2 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  padding: 16px;
  background-color: #054752;
  color: #FFF;
  border-radius: 16px;
  box-shadow: 0 0 2px rgba(0,0,0,.3);
  margin-left: 24px;
  margin-right: 24px;
}

.c2 strong {
  display: block;
  font-weight: 500;
  padding-bottom: 4px;
}

.c2 p {
  margin: 0;
  padding: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c2 .kirk-button {
  padding: 0 0 0 8px;
  min-height: 0;
  height: auto;
}

.c2 .kirk-button svg {
  margin: 0;
}

.c2.bubble-arrow--above::after {
  content: ' ';
  position: absolute;
  top: 100%;
  left: 50%;
  margin-left: -4px;
  width: 0;
  height: 0;
  border-left: 8px solid transparent;
  border-right: 8px solid transparent;
  border-top: 8px solid #054752;
}

.c2.bubble-arrow--below::after {
  content: ' ';
  position: absolute;
  bottom: 100%;
  left: 50%;
  margin-left: -4px;
  width: 0;
  height: 0;
  border-left: 8px solid transparent;
  border-right: 8px solid transparent;
  border-bottom: 8px solid #054752;
}

.c0 {
  position: relative;
}

.c0 .c1 {
  position: absolute;
  left: 0;
  right: 0;
  z-index: 2;
}

.c0 .c1.bubble-arrow--above {
  bottom: 100%;
}

.c0 .c1.bubble-arrow--below {
  top: 100%;
}

@media (hover:none),(hover:on-demand) {
  .c3.kirk-button-secondary:hover {
    background-color: #FFF;
  }
}

<div
  aria-live="polite"
  className="c0"
>
  <aside
    className="c1 c2 bubble-arrow--above"
    id="kirk-hint-1"
  >
    <p>
      <strong>
        Hint Title
      </strong>
      Hint Description
    </p>
    <button
      aria-controls="kirk-hint-1"
      className="kirk-button kirk-button-unstyled c3"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      title="Close"
      type="button"
    >
      <svg
        aria-hidden={true}
        className="kirk-icon c4"
        height={18}
        viewBox="0 0 24 24"
        width={18}
        xmlns="http://www.w3.org/2000/svg"
      >
        
        <path
          d="M19 5L5 19M19 19L5 5"
          fill="none"
          stroke="#FFF"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeMiterlimit="10"
          strokeWidth="2"
        />
      </svg>
    </button>
  </aside>
  <p
    aria-describedby="kirk-hint-1"
  >
    I have an hint.
  </p>
</div>
`;

exports[`Hint Default rendering (below) 1`] = `
.c4 {
  fill: #FFF;
}

.c4.kirk-icon-wrapper {
  display: inline-block;
  position: relative;
}

.c4.kirk-icon-shadowed {
  border-radius: 50%;
  box-shadow: 0 0 2px rgba(0,0,0,.3);
}

.c4 .kirk-icon-badge {
  position: absolute;
  bottom: 12px;
  left: 12px;
}

.c3 {
  position: relative;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding: 0 24px;
  border: 1px solid transparent;
  border-radius: 48px;
  font-size: 16px;
  line-height: 1;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  font-family: inherit;
  height: 48px;
  min-width: 48px;
  max-width: 100%;
  overflow: hidden;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: max-width 500ms ease-in, background-color 200ms ease-in;
  transition: max-width 500ms ease-in, background-color 200ms ease-in;
  white-space: nowrap;
}

.c3 svg,
.c3 img {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c3 svg {
  margin-right: 8px;
}

.c3.kirk-button-bubble svg {
  margin: 0;
}

.c3.kirk-itemChoice {
  border-radius: 0;
  overflow: visible;
  width: 100%;
}

.c3:hover,
.c3:active {
  outline: 0;
}

.c3.kirk-button-unstyled {
  background-color: transparent;
  border-radius: inherit;
  color: #00AFF5;
  padding: inherit;
  display: inherit;
  line-height: inherit;
  font-size: inherit;
  text-align: left;
  min-width: inherit;
  min-height: inherit;
  height: auto;
  white-space: normal;
}

.c3[disabled] {
  opacity: 0.5;
  cursor: default;
}

.c3.kirk-button-unstyled[disabled] {
  opacity: 1;
}

.c3.kirk-button-primary {
  background-color: #00AFF5;
  color: #FFF;
}

.c3.kirk-button-shadowed {
  border: 2px solid #FFF;
  box-shadow: 0 2px 4px rgba(0,0,0,.3);
  box-sizing: border-box;
}

.c3.kirk-button-primary:hover,
.c3.kirk-button-primary:focus,
.c3.kirk-button-primary:active {
  background-color: #008FC1;
}

.c3.kirk-button-secondary {
  background-color: #FFF;
  border-color: #DDD;
  color: #00AFF5;
}

.c3.kirk-button-secondary.kirk-button-shadowed {
  border: none;
}

.c3.kirk-button-secondary:hover,
.c3.kirk-button-secondary:focus,
.c3.kirk-button-secondary:active {
  background-color: #DDD;
}

.c3.kirk-button-tertiary {
  background-color: #FFF;
  border-color: transparent;
  color: #00AFF5;
}

.c3.kirk-button-tertiary:hover,
.c3.kirk-button-tertiary:focus,
.c3.kirk-button-tertiary:active {
  background-color: #DDD;
}

.c3.kirk-button-loading {
  background-color: transparent;
  max-width: initial;
  border: 0;
}

.c3.kirk-button-loading:hover,
.c3.kirk-button-loading:focus,
.c3.kirk-button-loading:active {
  background-color: transparent;
}

.c3.kirk-button-loading[disabled],
.c3.kirk-button-checked[disabled] {
  opacity: 1;
}

.c3 svg.kirk-button-loader {
  position: absolute;
  top: 0;
  left: 50%;
  margin-left: -24px;
}

.c3.kirk-button-checked {
  border: 0;
}

.c3.kirk-button-warning {
  background-color: #F53F5B;
  color: #FFF;
}

.c3.kirk-button-bubble {
  display: inline-block;
  text-align: center;
  padding: 0;
  line-height: 0;
  width: 48px;
  height: 48px;
}

.c3.kirk-button-bubble.kirk-button-shadowed {
  width: calc(48px - 2px * 2);
  height: calc(48px - 2px * 2);
}

.c3.kirk-button-facebook {
  background-color: #1877F2;
  color: #FFF;
}

.c3.kirk-button-vkontakte {
  background-color: #2787F5;
  color: #FFF;
}

.c2 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  padding: 16px;
  background-color: #054752;
  color: #FFF;
  border-radius: 16px;
  box-shadow: 0 0 2px rgba(0,0,0,.3);
  margin-left: 24px;
  margin-right: 24px;
}

.c2 strong {
  display: block;
  font-weight: 500;
  padding-bottom: 4px;
}

.c2 p {
  margin: 0;
  padding: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c2 .kirk-button {
  padding: 0 0 0 8px;
  min-height: 0;
  height: auto;
}

.c2 .kirk-button svg {
  margin: 0;
}

.c2.bubble-arrow--above::after {
  content: ' ';
  position: absolute;
  top: 100%;
  left: 50%;
  margin-left: -4px;
  width: 0;
  height: 0;
  border-left: 8px solid transparent;
  border-right: 8px solid transparent;
  border-top: 8px solid #054752;
}

.c2.bubble-arrow--below::after {
  content: ' ';
  position: absolute;
  bottom: 100%;
  left: 50%;
  margin-left: -4px;
  width: 0;
  height: 0;
  border-left: 8px solid transparent;
  border-right: 8px solid transparent;
  border-bottom: 8px solid #054752;
}

.c0 {
  position: relative;
}

.c0 .c1 {
  position: absolute;
  left: 0;
  right: 0;
  z-index: 2;
}

.c0 .c1.bubble-arrow--above {
  bottom: 100%;
}

.c0 .c1.bubble-arrow--below {
  top: 100%;
}

@media (hover:none),(hover:on-demand) {
  .c3.kirk-button-secondary:hover {
    background-color: #FFF;
  }
}

<div
  aria-live="polite"
  className="c0"
>
  <aside
    className="c1 c2 bubble-arrow--below"
    id="kirk-hint-2"
  >
    <p>
      <strong>
        Hint Title
      </strong>
      Hint Description
    </p>
    <button
      aria-controls="kirk-hint-2"
      className="kirk-button kirk-button-unstyled c3"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      title="Close"
      type="button"
    >
      <svg
        aria-hidden={true}
        className="kirk-icon c4"
        height={18}
        viewBox="0 0 24 24"
        width={18}
        xmlns="http://www.w3.org/2000/svg"
      >
        
        <path
          d="M19 5L5 19M19 19L5 5"
          fill="none"
          stroke="#FFF"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeMiterlimit="10"
          strokeWidth="2"
        />
      </svg>
    </button>
  </aside>
  <p
    aria-describedby="kirk-hint-2"
  >
    I have an hint.
  </p>
</div>
`;
